plugins/inputs/opcua_client/opcua_client.go

```diff
@@ -0,0 +1,408 @@
+		case "boolean", "byte", "short", "int", "uint", "uint16", "int16", "uint32", "int32", "float", "double", "string", "datetime", "number":
+			break
+		default:
+			return fmt.Errorf("invalid data type '%s' in '%s'", item.DataType, item.Name)
+		}
+
+		// build nodeid
+		o.Nodes = append(o.Nodes, BuildNodeID(item))
+
+		//parse NodeIds and NodeIds errors
+		nid, niderr := ua.ParseNodeID(o.Nodes[i])
+		// build NodeIds and Errors
+		o.NodeIDs = append(o.NodeIDs, nid)
+		o.NodeIDerror = append(o.NodeIDerror, niderr)
+		// Grow NodeData for later input
+		o.NodeData = append(o.NodeData, OPCData{})
+	}
+	return nil
+}
+
+// BuildNodeID build node ID from OPC tag
+func BuildNodeID(tag OPCTag) string {
+	return "ns=" + tag.Namespace + ";" + tag.IdentifierType + "=" + tag.Identifier
+}
+
+// Connect to a OPCUA Slave device
+func Connect(o *OpcUA) error {
+	u, err := url.Parse(o.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	switch u.Scheme {
+	case "opc.tcp":
+		o.state = Connecting
+
+		if o.client != nil {
+			o.client.CloseSession()
+		}
+
+		endpoints, err := opcua.GetEndpoints(o.Endpoint)
+		if err != nil {
+			return fmt.Errorf("Endpoint Error: %s", err)
+		}
+
+		ep := opcua.SelectEndpoint(endpoints, o.SecurityPolicy, ua.MessageSecurityModeFromString(o.SecurityMode))
+		if ep == nil {
+			return fmt.Errorf("Failed to find suitable endpoint")
+		}
+		opts := []opcua.Option{
+			opcua.SecurityPolicy(o.SecurityPolicy),
+			opcua.SecurityModeString(o.SecurityMode),
+			opcua.CertificateFile(o.Certificate),
+			opcua.PrivateKeyFile(o.PrivateKey),
+			opcua.AuthAnonymous(),
+			opcua.SecurityFromEndpoint(ep, ua.UserTokenTypeAnonymous),
+		}
+
+		o.client = opcua.NewClient(o.Endpoint, opts...)
+		if err := o.client.Connect(o.ctx); err != nil {
+			return fmt.Errorf("Error in Client Connection: %s", err)
+		}
+
+		regResp, err := o.client.RegisterNodes(&ua.RegisterNodesRequest{
+			NodesToRegister: o.NodeIDs,
+		})
+		if err != nil {
+			return fmt.Errorf("RegisterNodes failed: %v", err)
+		}
+
+		o.req = &ua.ReadRequest{
+			MaxAge:             2000,
+			NodesToRead:        readvalues(regResp.RegisteredNodeIDs),
+			TimestampsToReturn: ua.TimestampsToReturnBoth,
+		}
+
+		err = o.getData()
+		if err != nil {
+			return fmt.Errorf("Get Data Failed: %v", err)
+		}
+
+	default:
+		return fmt.Errorf("invalid opc.tcp endpoint")
+	}
+	return nil
+}
+
+func (o *OpcUA) getData() error {
+	resp, err := o.client.Read(o.req)
+	if err != nil {
+		o.ReadError++
+		return fmt.Errorf("RegisterNodes Read failed: %v", err)
+	}
+	o.ReadSuccess++
+	for i, d := range resp.Results {
+		if d.Status != ua.StatusOK {
+			return fmt.Errorf("Status not OK: %v", d.Status)
+		}
+		o.NodeData[i].TagName = o.NodeList[i].Name
+		if d.Value != nil {
+			o.NodeData[i].Value = d.Value.Value()
+			o.NodeData[i].DataType = d.Value.Type()
+		}
+		o.NodeData[i].Quality = d.Status
+		o.NodeData[i].TimeStamp = d.ServerTimestamp.String()
+		o.NodeData[i].Time = d.SourceTimestamp.String()
```

Comment:
string timestamps are notoriously difficult to work with. Not sure why you'd convert to a string here.